openedx/core/djangoapps/content/learning_sequences/api/tests/test_outlines.py

```diff
@@ -1279,6 +1280,39 @@ def test_private_course_outline(self):
                         'Sequences should be accessible to enrolled, staff users for a public_outline course'
 
 
+class UserPartitionGroupTestCase(OutlineProcessorTestCase):  # lint-amnesty, pylint: disable=missing-class-docstring
+    """Tests for user partitions that affect outlines (e.g. Enrollment Track)"""
+
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+    def test_roundtrip(self):
+        new_outline = CourseOutlineData(
+            course_key=self.course_key,
+            title="User Partition Test Course",
+            published_at=datetime(2021, 3, 26, tzinfo=timezone.utc),
+            published_version="8ebece4b69dd593d82fe2021",
+            sections=[
+                CourseSectionData(
+                    usage_key=self.course_key.make_usage_key('chapter', '0'),
+                    title="Section 0",
+                    user_partition_groups={
+                        50: frozenset([1, 2]),
```

Comment:
Nit: A real nit - but maybe add 2+ partition groups?

Reply:
Will do.

Reply:
You know, adding this test scenario led me to the interesting edge case of what happens when the XBlock defines a partition group setting on itself that has a Partition ID but no Groups. Like having a piece of content that says it's going to be partitioned by enrollment track, but then saying that no enrollment track group gets to see it. Right now the Django data model can't handle it at all–because there are no rows to create in that scenario (rows are key/value for this data, but there'd be no values in the empty case, so no rows).

I should check what the existing systems do. It might not be worth encoding at all if we're just going to handle it by ignoring it (which intuitively seems like the right thing?).

Reply:
Agree that it's worth checking other systems. But I also think that ignoring malformed user partitions sounds fine, as long as a log/error message is emitted when ignored.

Reply:
I went back and forth on this and I think I've settled on a slightly longer solution that echoes the pattern we've established for other content errors like this. Namely:

1. Learning Sequences outright rejects (throws an exception) when someone tries to specify something that it doesn't think is legal.
2. The outlines generation code in contentstore detects this edge case and puts an explicit ContentDataError into the system when it sees this kind of error, and then prunes it before it gets sent on to Learning Sequences.